CIP-0021/CIP-0021.md

```diff
@@ -0,0 +1,115 @@
+---
+CIP: 0021
+Title: Restrictions on transactions signed by hardware wallets
+Authors: Gabriel Kerekes <gabriel.kerekes@vacuumlabs.com>, Rafael Korbas <rafael.korbas@vacuumlabs.com>, Jan Mazak <jan.mazak@vacuumlabs.com>
+Status: Draft
+Type: Standards
+Created: 2021-06-15
+License: CC-BY-4.0
+---
+
+## Abstract
+
+This CIP describes all the restrictions posed when creating Cardano transactions which need to be signed by hardware wallets.
+
+## Motivation
+
+Due to certain limitations of HW wallets, especially very small memory and a limited set of data types supported by Ledger, HW wallets are not able to process all valid transactions which are supported by Cardano nodes.
+
+The limitations also result in an inability of HW wallets to see the whole transaction at once. Transaction data are streamed into HW wallets in small chunks and they compute a rolling hash of the transaction body which is signed at the end. Consequently, a HW wallet only provides witness signatures, and the transaction body which was signed has to be reconstructed by the client. We thus need a common transaction serialization format which will allow no ambiguity. In addition, the format must define ordering of map keys in such a way that it’s possible to check for duplicate keys by HW wallets.
+
+Several of the restrictions also stem from security or UX concerns e.g. the combination of pool registration certificates and withdrawals in a single transaction (see reasoning below).
+
+To ensure interoperability, SW wallets and other tools working with HW wallets should only use transactions which conform to the following rules.
+
+## Specification
+
+### Canonical CBOR serialization format
+
+Transactions must be serialized in line with suggestions from [Section 3.9 of CBOR specification RFC](https://datatracker.ietf.org/doc/html/rfc7049#section-3.9). In particular:
+
+- Integers must be as small as possible.
+- The expression of lengths in major types 2 through 5 must be as short as possible.
+- The keys in every map must be sorted from lowest value to highest.
+- Indefinite-length items must be made into definite-length items.
+
+See the RFC for details.
+
+### Transaction body
+
+Full support for some elements is not implemented yet (especially those related to scripts), but that is likely to change in the near future. The following restrictions are, however, unlikely to ever be lifted.
+
+**Unsupported entries**
+
+The transaction body entry `6 : update` must not be included.
+
+**Integers**
+
+While the CDDL specification usually does not limit the size of integers, HW wallets only support `int64` for signed integers and `uint64` for unsigned integers. Any integer value must fit in the appropriate type.
```

Comment:
Perhaps a link to the CDDL spec? But there are too many versions it does not deserve a definite article :D

Reply:
Since there are too many versions and even different files (shelley-ma, allonzo.cddl) adding a link might lead to it being outdated or it might lead to us having to update it regularly. I've updated it to say `While the Cardano CDDL specification` and IMHO it sounds ok.